// soya-next/lib/router/applyRedirect.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _regenerator = require('babel-runtime/regenerator');

var _regenerator2 = _interopRequireDefault(_regenerator);

var _typeof2 = require('babel-runtime/helpers/typeof');

var _typeof3 = _interopRequireDefault(_typeof2);

var _keys = require('babel-runtime/core-js/object/keys');

var _keys2 = _interopRequireDefault(_keys);

var _getIterator2 = require('babel-runtime/core-js/get-iterator');

var _getIterator3 = _interopRequireDefault(_getIterator2);

var _asyncToGenerator2 = require('babel-runtime/helpers/asyncToGenerator');

var _asyncToGenerator3 = _interopRequireDefault(_asyncToGenerator2);

var _extends2 = require('babel-runtime/helpers/extends');

var _extends3 = _interopRequireDefault(_extends2);

var _getPrototypeOf = require('babel-runtime/core-js/object/get-prototype-of');

var _getPrototypeOf2 = _interopRequireDefault(_getPrototypeOf);

var _classCallCheck2 = require('babel-runtime/helpers/classCallCheck');

var _classCallCheck3 = _interopRequireDefault(_classCallCheck2);

var _createClass2 = require('babel-runtime/helpers/createClass');

var _createClass3 = _interopRequireDefault(_createClass2);

var _possibleConstructorReturn2 = require('babel-runtime/helpers/possibleConstructorReturn');

var _possibleConstructorReturn3 = _interopRequireDefault(_possibleConstructorReturn2);

var _inherits2 = require('babel-runtime/helpers/inherits');

var _inherits3 = _interopRequireDefault(_inherits2);

var _jsxFileName = 'src/router/applyRedirect.js';

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _router = require('next/dist/lib/router/index.js');

var _router2 = _interopRequireDefault(_router);

var _querystring = require('querystring');

var _pathToRegexp = require('path-to-regexp');

var _pathToRegexp2 = _interopRequireDefault(_pathToRegexp);

var _getDisplayName = require('../utils/getDisplayName');

var _getDisplayName2 = _interopRequireDefault(_getDisplayName);

var _decodeParam = require('../utils/decodeParam');

var _decodeParam2 = _interopRequireDefault(_decodeParam);

var _ensureRedirect2 = require('../utils/ensureRedirect');

var _ensureRedirect3 = _interopRequireDefault(_ensureRedirect2);

var _parseRedirectionPath = require('../utils/parseRedirectionPath');

var _parseRedirectionPath2 = _interopRequireDefault(_parseRedirectionPath);

var _locale = require('../utils/locale');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

exports.default = function (Page) {
  var _class, _temp;

  return _temp = _class = function (_React$Component) {
    (0, _inherits3.default)(_class, _React$Component);

    function _class() {
      (0, _classCallCheck3.default)(this, _class);
      return (0, _possibleConstructorReturn3.default)(this, (_class.__proto__ || (0, _getPrototypeOf2.default)(_class)).apply(this, arguments));
    }

    (0, _createClass3.default)(_class, [{
      key: 'render',
      value: function render() {
        var props = (0, _extends3.default)({}, this.props);
        delete props.method;
        delete props.redirects;
        return _react2.default.createElement(Page, (0, _extends3.default)({}, props, {
          __source: {
            fileName: _jsxFileName,
            lineNumber: 58
          }
        }));
      }
    }], [{
      key: 'getInitialProps',
      value: function () {
        var _ref = (0, _asyncToGenerator3.default)(_regenerator2.default.mark(function _callee(ctx) {
          var _ref2, defaultLocale, locale, method, redirects, siteLocales, _iteratorNormalCompletion, _didIteratorError, _iteratorError, _iterator, _step, from, _ensureRedirect, redirectMethod, page, to, _ret, props;

          return _regenerator2.default.wrap(function _callee$(_context) {
            while (1) {
              switch (_context.prev = _context.next) {
                case 0:
                  _ref2 = ctx.req || window.__NEXT_DATA__.props, defaultLocale = _ref2.defaultLocale, locale = _ref2.locale, method = _ref2.method, redirects = _ref2.redirects, siteLocales = _ref2.siteLocales;

                  if (ctx.req) {
                    _context.next = 31;
                    break;
                  }

                  _iteratorNormalCompletion = true;
                  _didIteratorError = false;
                  _iteratorError = undefined;
                  _context.prev = 5;
                  _iterator = (0, _getIterator3.default)((0, _keys2.default)(redirects));

                case 7:
                  if (_iteratorNormalCompletion = (_step = _iterator.next()).done) {
                    _context.next = 17;
                    break;
                  }

                  from = _step.value;
                  _ensureRedirect = (0, _ensureRedirect3.default)(redirects[from]), redirectMethod = _ensureRedirect.method, page = _ensureRedirect.page, to = _ensureRedirect.to;

                  if (!(redirectMethod.toLowerCase() === method.toLowerCase())) {
                    _context.next = 14;
                    break;
                  }

                  _ret = function () {
                    var keys = [];
                    var regexp = (0, _pathToRegexp2.default)(from, keys);
                    var match = regexp.exec((0, _locale.trimPath)(ctx.asPath, defaultLocale, siteLocales));
                    if (match !== null) {
                      var localeSegment = (0, _locale.toPath)(locale, defaultLocale);
                      var params = keys.reduce(function (params, key, index) {
                        var param = match[index + 1];
                        // istanbul ignore else
                        if (param) {
                          params[key.name] = (0, _decodeParam2.default)(param);
                        }
                        return params;
                      }, {});
                      var redirectionPath = localeSegment + (0, _parseRedirectionPath2.default)(to, params);
                      _router2.default.push(page + '?' + (0, _querystring.stringify)(ctx.query), redirectionPath);
                      return {
                        v: {}
                      };
                    }
                  }();

                  if (!((typeof _ret === 'undefined' ? 'undefined' : (0, _typeof3.default)(_ret)) === "object")) {
                    _context.next = 14;
                    break;
                  }

                  return _context.abrupt('return', _ret.v);

                case 14:
                  _iteratorNormalCompletion = true;
                  _context.next = 7;
                  break;

                case 17:
                  _context.next = 23;
                  break;

                case 19:
                  _context.prev = 19;
                  _context.t0 = _context['catch'](5);
                  _didIteratorError = true;
                  _iteratorError = _context.t0;

                case 23:
                  _context.prev = 23;
                  _context.prev = 24;

                  if (!_iteratorNormalCompletion && _iterator.return) {
                    _iterator.return();
                  }

                case 26:
                  _context.prev = 26;

                  if (!_didIteratorError) {
                    _context.next = 29;
                    break;
                  }

                  throw _iteratorError;

                case 29:
                  return _context.finish(26);

                case 30:
                  return _context.finish(23);

                case 31:
                  _context.t1 = Page.getInitialProps;

                  if (!_context.t1) {
                    _context.next = 36;
                    break;
                  }

                  _context.next = 35;
                  return Page.getInitialProps(ctx);

                case 35:
                  _context.t1 = _context.sent;

                case 36:
                  props = _context.t1;
                  return _context.abrupt('return', (0, _extends3.default)({}, props, {
                    method: method,
                    redirects: redirects
                  }));

                case 38:
                case 'end':
                  return _context.stop();
              }
            }
          }, _callee, this, [[5, 19, 23, 31], [24,, 26, 30]]);
        }));

        function getInitialProps(_x) {
          return _ref.apply(this, arguments);
        }

        return getInitialProps;
      }()
    }]);
    return _class;
  }(_react2.default.Component), _class.displayName = (0, _getDisplayName2.default)('ApplyRedirect', Page), _temp;
};
