// soya-next/lib/components/LocaleLink.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _extends2 = require('babel-runtime/helpers/extends');

var _extends3 = _interopRequireDefault(_extends2);

var _objectWithoutProperties2 = require('babel-runtime/helpers/objectWithoutProperties');

var _objectWithoutProperties3 = _interopRequireDefault(_objectWithoutProperties2);

var _getPrototypeOf = require('babel-runtime/core-js/object/get-prototype-of');

var _getPrototypeOf2 = _interopRequireDefault(_getPrototypeOf);

var _classCallCheck2 = require('babel-runtime/helpers/classCallCheck');

var _classCallCheck3 = _interopRequireDefault(_classCallCheck2);

var _createClass2 = require('babel-runtime/helpers/createClass');

var _createClass3 = _interopRequireDefault(_createClass2);

var _possibleConstructorReturn2 = require('babel-runtime/helpers/possibleConstructorReturn');

var _possibleConstructorReturn3 = _interopRequireDefault(_possibleConstructorReturn2);

var _inherits2 = require('babel-runtime/helpers/inherits');

var _inherits3 = _interopRequireDefault(_inherits2);

var _jsxFileName = 'src/components/LocaleLink.js';

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _link = require('next/dist/lib/link.js');

var _link2 = _interopRequireDefault(_link);

var _redux = require('redux');

var _url = require('url');

var _withLocaleComponent = require('../i18n/withLocaleComponent');

var _withLocaleComponent2 = _interopRequireDefault(_withLocaleComponent);

var _withUrl = require('../router/withUrl');

var _withUrl2 = _interopRequireDefault(_withUrl);

var _PropTypes = require('../constants/PropTypes');

var _locale = require('../utils/locale');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var LocaleLink = function (_React$Component) {
  (0, _inherits3.default)(LocaleLink, _React$Component);

  function LocaleLink() {
    (0, _classCallCheck3.default)(this, LocaleLink);
    return (0, _possibleConstructorReturn3.default)(this, (LocaleLink.__proto__ || (0, _getPrototypeOf2.default)(LocaleLink)).apply(this, arguments));
  }

  (0, _createClass3.default)(LocaleLink, [{
    key: 'render',
    value: function render() {
      var _props = this.props,
          children = _props.children,
          defaultLocale = _props.defaultLocale,
          locale = _props.locale,
          url = _props.url,
          props = (0, _objectWithoutProperties3.default)(_props, ['children', 'defaultLocale', 'locale', 'url']);
      var as = props.as,
          href = props.href;

      delete props.siteLocales;

      var language = locale.language,
          country = locale.country;

      var localeSegment = (0, _locale.toPath)(locale, defaultLocale);

      href = href ? (0, _url.format)(href) : '';
      as = as ? (0, _url.format)(as) : '';
      var sep = href.indexOf('?') === -1 ? '?' : '&';

      return _react2.default.createElement(
        _link2.default,
        (0, _extends3.default)({}, props, {
          as: localeSegment + (0, _url.resolve)(url.pathname, as || href),
          href: '' + (0, _url.resolve)(url.pathname, href) + sep + 'locale=' + language + '-' + country,
          __source: {
            fileName: _jsxFileName,
            lineNumber: 38
          }
        }),
        children
      );
    }
  }]);
  return LocaleLink;
}(_react2.default.Component);

LocaleLink.propTypes = {
  children: _propTypes2.default.element.isRequired,
  defaultLocale: _propTypes2.default.string.isRequired,
  locale: _PropTypes.localeShape.isRequired,
  url: _PropTypes.urlShape.isRequired
};
exports.default = (0, _redux.compose)(_withLocaleComponent2.default, _withUrl2.default)(LocaleLink);
