// soya-next/lib/components/SoyaProvider.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _getPrototypeOf = require('babel-runtime/core-js/object/get-prototype-of');

var _getPrototypeOf2 = _interopRequireDefault(_getPrototypeOf);

var _classCallCheck2 = require('babel-runtime/helpers/classCallCheck');

var _classCallCheck3 = _interopRequireDefault(_classCallCheck2);

var _createClass2 = require('babel-runtime/helpers/createClass');

var _createClass3 = _interopRequireDefault(_createClass2);

var _possibleConstructorReturn2 = require('babel-runtime/helpers/possibleConstructorReturn');

var _possibleConstructorReturn3 = _interopRequireDefault(_possibleConstructorReturn2);

var _inherits2 = require('babel-runtime/helpers/inherits');

var _inherits3 = _interopRequireDefault(_inherits2);

var _jsxFileName = 'src/components/SoyaProvider.js';

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _reactCookie = require('react-cookie');

var _reactRedux = require('react-redux');

var _PropTypes = require('react-redux/lib/utils/PropTypes');

var _PropTypes2 = require('../constants/PropTypes');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var SoyaProvider = function (_React$Component) {
  (0, _inherits3.default)(SoyaProvider, _React$Component);

  function SoyaProvider() {
    (0, _classCallCheck3.default)(this, SoyaProvider);
    return (0, _possibleConstructorReturn3.default)(this, (SoyaProvider.__proto__ || (0, _getPrototypeOf2.default)(SoyaProvider)).apply(this, arguments));
  }

  (0, _createClass3.default)(SoyaProvider, [{
    key: 'getChildContext',
    value: function getChildContext() {
      return {
        defaultLocale: this.props.defaultLocale,
        siteLocales: this.props.siteLocales,
        locale: this.props.locale,
        url: this.props.url
      };
    }
  }, {
    key: 'render',
    value: function render() {
      return _react2.default.createElement(
        _reactCookie.CookiesProvider,
        { cookies: this.props.cookies, __source: {
            fileName: _jsxFileName,
            lineNumber: 37
          }
        },
        _react2.default.createElement(
          _reactRedux.Provider,
          { store: this.props.store, __source: {
              fileName: _jsxFileName,
              lineNumber: 38
            }
          },
          _react2.default.Children.only(this.props.children)
        )
      );
    }
  }]);
  return SoyaProvider;
}(_react2.default.Component);

SoyaProvider.propTypes = {
  store: _PropTypes.storeShape.isRequired,
  cookies: _propTypes2.default.instanceOf(_reactCookie.Cookies).isRequired,
  defaultLocale: _propTypes2.default.string,
  siteLocales: _propTypes2.default.arrayOf(_propTypes2.default.string.isRequired),
  locale: _PropTypes2.localeShape,
  url: _PropTypes2.urlShape,
  children: _propTypes2.default.element.isRequired
};
SoyaProvider.childContextTypes = {
  defaultLocale: _propTypes2.default.string,
  siteLocales: _propTypes2.default.arrayOf(_propTypes2.default.string.isRequired),
  locale: _PropTypes2.localeShape,
  url: _PropTypes2.urlShape
};
exports.default = SoyaProvider;
