// soya-next/lib/redux/applyReducers.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _extends2 = require('babel-runtime/helpers/extends');

var _extends3 = _interopRequireDefault(_extends2);

var _getPrototypeOf = require('babel-runtime/core-js/object/get-prototype-of');

var _getPrototypeOf2 = _interopRequireDefault(_getPrototypeOf);

var _regenerator = require('babel-runtime/regenerator');

var _regenerator2 = _interopRequireDefault(_regenerator);

var _asyncToGenerator2 = require('babel-runtime/helpers/asyncToGenerator');

var _asyncToGenerator3 = _interopRequireDefault(_asyncToGenerator2);

var _classCallCheck2 = require('babel-runtime/helpers/classCallCheck');

var _classCallCheck3 = _interopRequireDefault(_classCallCheck2);

var _possibleConstructorReturn2 = require('babel-runtime/helpers/possibleConstructorReturn');

var _possibleConstructorReturn3 = _interopRequireDefault(_possibleConstructorReturn2);

var _createClass2 = require('babel-runtime/helpers/createClass');

var _createClass3 = _interopRequireDefault(_createClass2);

var _inherits2 = require('babel-runtime/helpers/inherits');

var _inherits3 = _interopRequireDefault(_inherits2);

var _jsxFileName = 'src/redux/applyReducers.js';

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _PropTypes = require('react-redux/lib/utils/PropTypes');

var _getDisplayName = require('../utils/getDisplayName');

var _getDisplayName2 = _interopRequireDefault(_getDisplayName);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

exports.default = function (reducers) {
  return function (Component) {
    if (typeof reducers === 'undefined') {
      return Component;
    }

    var ApplyReducers = function (_React$Component) {
      (0, _inherits3.default)(ApplyReducers, _React$Component);
      (0, _createClass3.default)(ApplyReducers, null, [{
        key: 'getInitialProps',
        value: function () {
          var _ref = (0, _asyncToGenerator3.default)(_regenerator2.default.mark(function _callee(ctx) {
            return _regenerator2.default.wrap(function _callee$(_context) {
              while (1) {
                switch (_context.prev = _context.next) {
                  case 0:
                    if (ctx.store.soya) {
                      _context.next = 2;
                      break;
                    }

                    throw new Error('applyReducers must be used with Soya\'s redux enhancer');

                  case 2:
                    ctx.store.addReducer(reducers);
                    _context.t0 = Component.getInitialProps;

                    if (!_context.t0) {
                      _context.next = 8;
                      break;
                    }

                    _context.next = 7;
                    return Component.getInitialProps(ctx);

                  case 7:
                    _context.t0 = _context.sent;

                  case 8:
                    return _context.abrupt('return', _context.t0);

                  case 9:
                  case 'end':
                    return _context.stop();
                }
              }
            }, _callee, this);
          }));

          function getInitialProps(_x) {
            return _ref.apply(this, arguments);
          }

          return getInitialProps;
        }()
      }]);

      function ApplyReducers(props, context) {
        (0, _classCallCheck3.default)(this, ApplyReducers);

        var _this = (0, _possibleConstructorReturn3.default)(this, (ApplyReducers.__proto__ || (0, _getPrototypeOf2.default)(ApplyReducers)).call(this, props, context));

        _this.store = props.store || context.store;
        if (!_this.store.soya) {
          throw new Error('applyReducers must be used with Soya\'s redux enhancer');
        }
        _this.store.addReducer(reducers);
        return _this;
      }

      (0, _createClass3.default)(ApplyReducers, [{
        key: 'render',
        value: function render() {
          return _react2.default.createElement(Component, (0, _extends3.default)({}, this.props, {
            __source: {
              fileName: _jsxFileName,
              lineNumber: 35
            }
          }));
        }
      }]);
      return ApplyReducers;
    }(_react2.default.Component);

    ApplyReducers.displayName = (0, _getDisplayName2.default)('ApplyReducers', Component);
    ApplyReducers.contextTypes = {
      store: _PropTypes.storeShape.isRequired
    };

    return ApplyReducers;
  };
};
