// soya-next/lib/redux/withStore.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _objectWithoutProperties2 = require('babel-runtime/helpers/objectWithoutProperties');

var _objectWithoutProperties3 = _interopRequireDefault(_objectWithoutProperties2);

var _getPrototypeOf = require('babel-runtime/core-js/object/get-prototype-of');

var _getPrototypeOf2 = _interopRequireDefault(_getPrototypeOf);

var _regenerator = require('babel-runtime/regenerator');

var _regenerator2 = _interopRequireDefault(_regenerator);

var _extends2 = require('babel-runtime/helpers/extends');

var _extends3 = _interopRequireDefault(_extends2);

var _asyncToGenerator2 = require('babel-runtime/helpers/asyncToGenerator');

var _asyncToGenerator3 = _interopRequireDefault(_asyncToGenerator2);

var _classCallCheck2 = require('babel-runtime/helpers/classCallCheck');

var _classCallCheck3 = _interopRequireDefault(_classCallCheck2);

var _possibleConstructorReturn2 = require('babel-runtime/helpers/possibleConstructorReturn');

var _possibleConstructorReturn3 = _interopRequireDefault(_possibleConstructorReturn2);

var _createClass2 = require('babel-runtime/helpers/createClass');

var _createClass3 = _interopRequireDefault(_createClass2);

var _inherits2 = require('babel-runtime/helpers/inherits');

var _inherits3 = _interopRequireDefault(_inherits2);

var _jsxFileName = 'src/redux/withStore.js';

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _PropTypes = require('react-redux/lib/utils/PropTypes');

var _getDisplayName = require('../utils/getDisplayName');

var _getDisplayName2 = _interopRequireDefault(_getDisplayName);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

exports.default = function (configureStore) {
  return function (Page) {
    var configureStoreIfNeeded = function configureStoreIfNeeded(preloadedState, extraArgument) {
      if (typeof window === 'undefined') {
        return configureStore(preloadedState, extraArgument);
      }
      if (typeof window.store === 'undefined') {
        window.store = configureStore(preloadedState, extraArgument);
      }
      return window.store;
    };

    var WithStore = function (_React$Component) {
      (0, _inherits3.default)(WithStore, _React$Component);
      (0, _createClass3.default)(WithStore, null, [{
        key: 'getInitialProps',
        value: function () {
          var _ref = (0, _asyncToGenerator3.default)(_regenerator2.default.mark(function _callee(ctx) {
            var store, props;
            return _regenerator2.default.wrap(function _callee$(_context) {
              while (1) {
                switch (_context.prev = _context.next) {
                  case 0:
                    store = configureStoreIfNeeded(undefined, { cookies: ctx.cookies });
                    _context.t0 = Page.getInitialProps;

                    if (!_context.t0) {
                      _context.next = 6;
                      break;
                    }

                    _context.next = 5;
                    return Page.getInitialProps((0, _extends3.default)({}, ctx, { store: store }));

                  case 5:
                    _context.t0 = _context.sent;

                  case 6:
                    props = _context.t0;
                    return _context.abrupt('return', (0, _extends3.default)({}, props, {
                      preloadedState: store.getState(),
                      store: store
                    }));

                  case 8:
                  case 'end':
                    return _context.stop();
                }
              }
            }, _callee, this);
          }));

          function getInitialProps(_x) {
            return _ref.apply(this, arguments);
          }

          return getInitialProps;
        }()
      }]);

      function WithStore(props) {
        (0, _classCallCheck3.default)(this, WithStore);

        var _this = (0, _possibleConstructorReturn3.default)(this, (WithStore.__proto__ || (0, _getPrototypeOf2.default)(WithStore)).call(this, props));

        _this.store = props.store.dispatch ? props.store : configureStoreIfNeeded(props.preloadedState, { cookies: props.cookies });
        return _this;
      }

      (0, _createClass3.default)(WithStore, [{
        key: 'render',
        value: function render() {
          var props = (0, _objectWithoutProperties3.default)(this.props, []);

          delete props.preloadedState;
          delete props.store;

          return _react2.default.createElement(Page, (0, _extends3.default)({}, props, { store: this.store, __source: {
              fileName: _jsxFileName,
              lineNumber: 49
            }
          }));
        }
      }]);
      return WithStore;
    }(_react2.default.Component);

    WithStore.displayName = (0, _getDisplayName2.default)('WithStore', Page);
    WithStore.propTypes = {
      preloadedState: _propTypes2.default.object.isRequired,
      store: _propTypes2.default.oneOfType([_propTypes2.default.object, _PropTypes.storeShape]).isRequired
    };


    return WithStore;
  };
};
