// soya-next/lib/pages/createPageFactory.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _extends2 = require('babel-runtime/helpers/extends');

var _extends3 = _interopRequireDefault(_extends2);

var _objectWithoutProperties2 = require('babel-runtime/helpers/objectWithoutProperties');

var _objectWithoutProperties3 = _interopRequireDefault(_objectWithoutProperties2);

var _getPrototypeOf = require('babel-runtime/core-js/object/get-prototype-of');

var _getPrototypeOf2 = _interopRequireDefault(_getPrototypeOf);

var _classCallCheck2 = require('babel-runtime/helpers/classCallCheck');

var _classCallCheck3 = _interopRequireDefault(_classCallCheck2);

var _createClass2 = require('babel-runtime/helpers/createClass');

var _createClass3 = _interopRequireDefault(_createClass2);

var _possibleConstructorReturn2 = require('babel-runtime/helpers/possibleConstructorReturn');

var _possibleConstructorReturn3 = _interopRequireDefault(_possibleConstructorReturn2);

var _inherits2 = require('babel-runtime/helpers/inherits');

var _inherits3 = _interopRequireDefault(_inherits2);

var _jsxFileName = 'src/pages/createPageFactory.js';

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _hoistNonReactStatics = require('hoist-non-react-statics');

var _hoistNonReactStatics2 = _interopRequireDefault(_hoistNonReactStatics);

var _redux = require('redux');

var _reactRedux = require('react-redux');

var _PropTypes = require('react-redux/lib/utils/PropTypes');

var _reactCookie = require('react-cookie');

var _PropTypes2 = require('../constants/PropTypes');

var _SoyaProvider = require('../components/SoyaProvider');

var _SoyaProvider2 = _interopRequireDefault(_SoyaProvider);

var _applyRedirect = require('../router/applyRedirect');

var _applyRedirect2 = _interopRequireDefault(_applyRedirect);

var _applyReducers = require('../redux/applyReducers');

var _applyReducers2 = _interopRequireDefault(_applyReducers);

var _withCookiesPage = require('../cookies/withCookiesPage');

var _withCookiesPage2 = _interopRequireDefault(_withCookiesPage);

var _withLocalePage = require('../i18n/withLocalePage');

var _withLocalePage2 = _interopRequireDefault(_withLocalePage);

var _withStore = require('../redux/withStore');

var _withStore2 = _interopRequireDefault(_withStore);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

exports.default = function (configureStore) {
  return function () {
    for (var _len = arguments.length, connectArgs = Array(_len), _key = 0; _key < _len; _key++) {
      connectArgs[_key] = arguments[_key];
    }

    return function (Page, reducers) {
      var EnhancedPage = (0, _redux.compose)((0, _applyReducers2.default)(reducers), _reactRedux.connect.apply(undefined, connectArgs))(Page);

      var SoyaPage = function (_React$Component) {
        (0, _inherits3.default)(SoyaPage, _React$Component);

        function SoyaPage() {
          (0, _classCallCheck3.default)(this, SoyaPage);
          return (0, _possibleConstructorReturn3.default)(this, (SoyaPage.__proto__ || (0, _getPrototypeOf2.default)(SoyaPage)).apply(this, arguments));
        }

        (0, _createClass3.default)(SoyaPage, [{
          key: 'render',
          value: function render() {
            var _props = this.props,
                store = _props.store,
                props = (0, _objectWithoutProperties3.default)(_props, ['store']);

            return _react2.default.createElement(
              _SoyaProvider2.default,
              {
                cookies: this.props.cookies,
                locale: this.props.locale,
                defaultLocale: this.props.defaultLocale,
                siteLocales: this.props.siteLocales,
                store: store,
                url: this.props.url,
                __source: {
                  fileName: _jsxFileName,
                  lineNumber: 35
                }
              },
              _react2.default.createElement(EnhancedPage, (0, _extends3.default)({}, props, {
                __source: {
                  fileName: _jsxFileName,
                  lineNumber: 43
                }
              }))
            );
          }
        }]);
        return SoyaPage;
      }(_react2.default.Component);

      SoyaPage.propTypes = {
        cookies: _propTypes2.default.instanceOf(_reactCookie.Cookies).isRequired,
        store: _PropTypes.storeShape.isRequired,
        defaultLocale: _propTypes2.default.string,
        siteLocales: _propTypes2.default.arrayOf(_propTypes2.default.string.isRequired),
        locale: _PropTypes2.localeShape,
        url: _PropTypes2.urlShape.isRequired
      };


      return (0, _redux.compose)(_withLocalePage2.default, _applyRedirect2.default, _withCookiesPage2.default, (0, _withStore2.default)(configureStore))((0, _hoistNonReactStatics2.default)(SoyaPage, EnhancedPage));
    };
  };
};
